src/hooks/usePermitAllowance.ts

```diff
@@ -0,0 +1,88 @@
+import {
+  AllowanceData,
+  AllowanceProvider,
+  AllowanceTransfer,
+  MaxAllowanceTransferAmount,
+  PERMIT2_ADDRESS,
+  PermitSingle,
+} from '@uniswap/permit2-sdk'
+import { Token } from '@uniswap/sdk-core'
+import { useWeb3React } from '@web3-react/core'
+import ms from 'ms.macro'
+import { useCallback, useEffect, useMemo, useState } from 'react'
+
+const PERMIT_EXPIRATION = ms`30d`
+const PERMIT_SIG_EXPIRATION = ms`30m`
+
+function toDeadline(expiration: number): number {
+  return Math.floor((Date.now() + expiration) / 1000)
+}
+
+export function usePermitAllowance(token?: Token, spender?: string) {
+  const { account, provider } = useWeb3React()
+  const allowanceProvider = useMemo(() => provider && new AllowanceProvider(provider, PERMIT2_ADDRESS), [provider])
+  const [allowanceData, setAllowanceData] = useState<AllowanceData>()
+
+  useEffect(() => {
+    if (!account || !token || !spender) return
+
+    allowanceProvider
+      ?.getAllowanceData(token.address, account, spender)
+      .then((data) => {
+        if (stale) return
+        setAllowanceData(data)
+      })
+      .catch((e) => {
+        console.warn(`Failed to fetch allowance data: ${e}`)
+      })
+
+    let stale = false
+    return () => {
+      stale = true
+    }
+  }, [account, allowanceProvider, spender, token])
+
+  return allowanceData
+}
+
+interface Permit extends PermitSingle {
+  sigDeadline: number
+}
+
+export interface PermitSignature extends Permit {
+  signature: string
+}
+
+export function usePermitAllowanceCallback(
+  token?: Token,
+  spender?: string,
+  nonce?: number,
+  onPermitSignature?: (signature: PermitSignature) => void
+) {
+  const { account, chainId, provider } = useWeb3React()
+
+  return useCallback(async () => {
+    try {
+      if (!chainId || !provider || !token || !spender || nonce === undefined) throw new Error('missing input')
+
+      const permit: Permit = {
+        details: {
+          token: token.address,
+          amount: MaxAllowanceTransferAmount,
+          expiration: toDeadline(PERMIT_EXPIRATION),
+          nonce,
+        },
+        spender,
+        sigDeadline: toDeadline(PERMIT_SIG_EXPIRATION),
```

Comment:
for my own edification, what's the diff here btwn the two expiration constants and why?

Reply:
SigDeadline describes when the signature is no longer valid. Expiration describes when the _approval_ is no longer valid. So if we submit our permit within 30 minutes then we have unlimited approval for 30 days after that